---
title: "RNA-seq data analysis with DESeq2"
output: html_document
bibliography: bib.bib
---


# RNA-seq data analysis with DESeq2

Presented By Petr Smirnov^[University of Toronto, CA]

Heavily Adapted From:
    Michael I. Love^[UNC-Chapel Hill, NC, US],
    Simon Anders^[ZMBH Heidelberg, Germany],
    Wolfgang Huber^[EMBL Heidelberg, Germany]
Last modified: Jan 31, 2019.

## Overview

### Description

In this workshop, we will give a quick overview of the most useful
functions in the DESeq2 package, and a basic RNA-seq analysis. We will
cover: how to quantify transcript expression from FASTQ files using
Kallisto, import quantification from Kallisto with tximport,
generate plots for quality control and exploratory data analysis EDA,
perform differential expression (DE) (also using apeglm), 
overlap with other annotation data (using AnnotationHub),
and build reports (using ReportingTools and Glimma). The workshop is designed
to be a lab with plenty of time for questions throughout the lab. 

### Pre-requisites

* Basic knowledge of R syntax

Non-essential background reading:

* DESeq2 paper: <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4302049/>
* tximport paper: <https://f1000research.com/articles/4-1521/v2>
* apeglm paper: <https://www.biorxiv.org/content/early/2018/04/17/303255>

### Participation

Students will participate by following along an Rmarkdown document,
and asking questions throughout the workshop.

### _R_ / _Bioconductor_ packages used

* DESeq2
* tximport
* apeglm
* AnnotationHub
* ReportingTools
* Glimma

### Time outline

| Activity                      | Time |
|:------------------------------|:-----|
| Overview of packages          | 20m  |
| Quantification and import     | 20m  |
| EDA and DE                    | 20m  |
| Downstream analysis & reports | 20m  |
| Additional questions          | 20m  |

### Workshop goals and objectives

Learning goals

* Visually assess quality of RNA-seq data 
* Perform basic differential analysis of RNA-seq data 

Learning objectives

* Learn how transcript expression is quantified from FASTQ files
* Import quantification into R
* Perform quality control and exploratory data analysis
* Perform differential expression
* Build dynamic reports

## Preparing data for *DESeq2*

### Experimental data

The data used in this workflow comes from the *airway* package that
summarizes an RNA-seq experiment wherein airway smooth muscle cells
were treated with dexamethasone, a synthetic glucocorticoid steroid
with anti-inflammatory effects [@Himes2014RNASeq]. Glucocorticoids
are used, for example, by people with asthma to reduce inflammation of
the airways. In the experiment, four primary human airway smooth
muscle cell lines were treated with 1 micromolar dexamethasone for 18
hours. For each of the four cell lines, we have a treated and an
untreated sample. For more description of the experiment see the
[PubMed entry 24926665](http://www.ncbi.nlm.nih.gov/pubmed/24926665)
and for raw data see the
[GEO entry GSE52778](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE52778).

### Modeling count data

As input, the count-based statistical methods, such as *DESeq2*
[@Love2014Moderated], *edgeR* [@Robinson2009EdgeR], *limma* with the
voom method [@Law2014Voom], *DSS* [@Wu2013New], *EBSeq*
[@Leng2013EBSeq] and *baySeq* [@Hardcastle2010BaySeq], expect input
data as obtained, e.g., from RNA-seq or another high-throughput
sequencing experiment, in the form of a matrix of counts.  The value
in the *i*-th row and the *j*-th column of the matrix tells how many
reads (or fragments, for paired-end RNA-seq) have been assigned to
gene *i* in sample *j*. Analogously, for other types of assays, the
rows of the matrix might correspond e.g., to binding regions (with
ChIP-Seq), species of bacteria (with metagenomic datasets), or peptide
sequences (with quantitative mass spectrometry).

The values in the matrix should be counts of sequencing
reads/fragments. This is important for the statistical models used by
*DESeq2* and *edgeR* to hold, as only counts allow assessing the
measurement precision correctly. It is important to not provide counts
that were pre-normalized for sequencing depth (also called library
size), as the statistical model is most powerful when applied to
un-normalized counts and is designed to account for library size
differences internally.

### Transcript abundances

In this workflow, we will show how to use transcript abundances as
quantified by the [Kallisto](https://pachterlab.github.io/kallisto/)
[@Bray2016Near] software package. *Kallisto* and other methods, such
as [Sailfish](http://www.cs.cmu.edu/~ckingsf/software/sailfish/)
[@Patro2014Sailfish],
[Salmon](https://combine-lab.github.io/salmon/) [@Patro2017Salmon],
or [RSEM](http://deweylab.github.io/RSEM/) [@Li2011RSEM], estimate the
relative abundances of all (known, annotated) transcripts without
aligning reads. Because estimating the abundance of the transcripts
involves an inference step, the counts are *estimated*. Most methods
either use a statistical framework called Estimation-Maximization or
Bayesian techniques to estimate the abundances and counts.  Following
quantification, we will use the *tximport* [@Soneson2015Differential]
package for assembling estimated count and offset matrices for use
with Bioconductor differential gene expression packages.

The advantages of using the transcript abundance quantifiers in
conjunction with *tximport* to produce gene-level count matrices and
normalizing offsets, are: 

1. this approach corrects for any potential changes in gene length
across samples (e.g. from differential isoform usage)
[@Trapnell2013Differential] 
2. some of these methods are substantially faster and require less
memory and less disk usage compared to alignment-based methods 
3. it is possible to avoid discarding those fragments that can align
to multiple genes with homologous sequence [@Robert2015Errors]. 

Note that transcript abundance quantifiers skip the generation of
large files which store read alignments (SAM or BAM files), instead
producing smaller files which store estimated abundances, counts and
effective lengths per transcript. For more details, see the manuscript
describing this approach [@Soneson2015Differential] and the *tximport*
package vignette for software details.



### *kallisto* quantification

We begin by providing *kallisto* with the sequence of all of the
reference transcripts, which we will call the *reference
transcriptome*. We recommend to use the GENCODE human
transcripts, which can be downloaded from the
[GENCODE website](https://www.gencodegenes.org/). Kalliso needs to 
once create an index of these transcripts, which can be reused as 
long as you want to quantify using the same transcriptome.
On the command line, creating the transcriptome index looks like:

```
kallisto index transcriptome/gencode.v29.transcripts.fa.gz -i transcriptome/gencodev29.kalliso.v45.idx
```

The `v45` refers to the version of *kallisto* that was used, and is
useful to put into the index name.

To quantify an individual sample, for example `SRR1039508`, the following command
can be used:

```
kallisto quant -i kallisto-45/transcriptome/gencodev29.kalliso.v45.idx\
  -o output/SRR1039508 \
  -t 3 \
  fastq/SRR1039508_1.fastq\
  fastq/SRR1039508_2.fastq
```

In simple English, this command says to "quantify a sample using this
transcriptome index, use this output directory, using 3 of the computer
processors, and here are the first and second
read files." The output directory will be
created if it doesn't exist, though if earlier parts of the path do
not exist, it will give an error. A single sample of human RNA-seq
usually takes ~5 minutes.

Rather than writing the above command on the command line multiple
times for each sample, it is possible to loop over files using a
bash loop, which is what we do below:


```{bash, eval=FALSE}
files=`cat samples.txt`
for file in $files
do
kallisto-45/kallisto quant -i kallisto-45/transcriptome/gencodev29.kalliso.v45.idx -o output/"$file" -t 3  --plaintext fastq/"$file"_1.fastq fastq/"$file"_2.fastq
done

```


## Importing into R with *tximport*

### Specifying file locations

Following quantification, we can use *tximport* to import the data
into R and perform statistical analysis using Bioconductor packages.

The identifiers used here are the *SRA* identifiers from the 
[Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra). 
We need to create a named vector pointing to the quantification
files. It is recommended to create a sample table mapping each file
to the sample and condition of the sequence run. We precreated a table
for this subset of the *airways* data. 


```{r}
mydir <- getwd()
samples <- read.csv(file.path(mydir,"samples.csv"), header=TRUE)
samples
files <- file.path(mydir, "output", samples$SampleName, "abundance.tsv")
names(files) <- samples$SampleName
all(file.exists(files))
```

### Mapping transcripts to genes

Transcripts need to be associated with gene IDs for gene-level
summarization. We therefore will construct a *data.frame* called
`tx2gene` with two columns: 1) transcript ID and 2) gene ID. The
column names do not matter but this column order must be used. The
transcript ID must be the same one used in the abundance files. This
can most easily be accomplished by downloading the GTF file at the
same time that the transcriptome FASTA is downloaded, and generating
`tx2gene` from the GTF file using Bioconductor's *TxDb*
infrastructure.

Generating a *TxDb* from a GTF file can be easily accomplished with
the *makeTxDbFromGFF* function, but this step requires a few minutes of
waiting, and a large file, so we precomputed the results:
```
txdb <- makeTxDbFromGFF("transcriptome/gencode.v29.annotation.gtf.gz")
saveDb(txdb, "transcriptome/gencode.v29.annotation.txdb")
```

Creating the `tx2gene` *data.frame* can be accomplished by calling the
*select* function from the *AnnotationDbi* package on a *TxDb* object.
The following code could be used to construct such a table:

```{r}
library(AnnotationDbi) 
txdb <- loadDb("transcriptome/gencode.v29.annotation.txdb")
k <- keys(txdb, keytype="TXNAME")
tx2gene <- select(txdb, k, "GENEID", "TXNAME")
```

### *tximport* command

Finally the following line of code imports *Kallisto* transcript
quantifications into R, collapsing to the gene level using the
information in `tx2gene`.

```{r}
library("tximport")
library("jsonlite")
library("readr")
txi <- tximport(files, type="kallisto", tx2gene=tx2gene, ignoreAfterBar = TRUE)
```

The `txi` object is simply a list of matrices (and one character
vector):

```{r}
names(txi)
txi$counts[1:3,1:3]
txi$length[1:3,1:3]
txi$abundance[1:3,1:3]
txi$countsFromAbundance
```

Now that we have used the *tximport* package to load our files into R,
we can start our analysis. For the purpose of this tutorial we are using
the methods implemented in the DESeq2 package, so its time to load it in, 
and create a DESeqDataSet to use with the functions in the package. Here, 
we provide the sample table with the group annotations for our cells, and 
for now we provide a dummy design, assuming every cell is in the same group.


```{r}
library("DESeq2")
dds <- DESeqDataSetFromTximport(txi, colData = samples, design = ~1)
```

## Exploratory data analysis

### Simple EDA
In what follows, we will be comparing the gene expression of the dex treated vs
untreated cells. To ensure we know how to interpret the directions, we want to 
specify that `untrt` is the reference level for the dex variable, using R's relevel
command: 

```{r}
dds$dex <- relevel(dds$dex, "untrt")
dds$dex
```

We can quickly check the millions of fragments that uniquely aligned
to the genes (the second argument of *round* tells how many decimal
points to keep).

```{r}
round( colSums(assay(dds)) / 1e6, 1 )
```

We can inspect the information about the samples, by pulling out the
`colData` slot of the *SummarizedExperiment*:

```{r}
colData(dds)
table(dds$cell)
table(dds$dex)
```

<!-- If we had not already loaded *DESeq2*, we would do this, and then -->
<!-- create a *DESeqDataSet*. We want to control for the cell line, while -->
<!-- testing for differences across dexamethasone treatment, so we use a -->
<!-- design of `~ cell + dex`: -->

<!-- ```{r} -->
<!-- library("DESeq2") -->
<!-- dds <- DESeqDataSet(airway, design = ~ cell + dex) -->
<!-- ``` -->

We will perform a minimal filtering to reduce the size of the
dataset. We do not need to retain genes if they do not have a count of
5 or more for 4 or more samples as these genes will have no
statistical power to detect differences, and no information to compute
distances between samples.

```{r}
keep <- rowSums(counts(dds) >= 5) >= 4
table(keep)
dds <- dds[keep,]
```

Some very basic exploratory analysis is to examine a boxplot of the
counts for each sample. We will take the logarithm so that large
counts do not dominate the boxplot:

```{r}
boxplot(log10(counts(dds)+1))
```

The main function in *DESeq2* involves computation of *size factors*
which normalize for differences in sequencing depth among samples. We
can also compute these size factors manually, so that the *normalized
counts* are available for plotting:

```{r}
dds <- estimateSizeFactors(dds)
boxplot(log10(counts(dds,normalized=TRUE)+1))
```

### Data transformation for EDA

Taking the logarithm of counts plus a pseudocount of 1 is a common
transformation, but it tends to inflate the sampling variance of low
counts such that it is even larger than biological variation across
groups of samples. In *DESeq2* we therefore provide transformations
which produce log-scale data such that the systematic trends have been
removed. Our recommended transformation is the variance-stabilizing
transformation, or VST, and it can be called with the *vst* function:

```{r}
vsd <- vst(dds)
class(vsd)
```

This function does not return a *DESeqDataSet*, because it does not
return counts, but instead continuous values (on the log2 scale).
We can access the transformed data with *assay*:

```{r}
assay(vsd)[1:3,1:3]
```

### Why VST?

The reason we should usually perform a Variance Stabalizing transformation 
is because count data has the property that the variance of expression 
depends highly on the average expression. If we then apply a method like PCA,
it will be biased by the higher variance in lower expressed genes, and will 
overestimate the importance in the differences observed in those genes. 

We can visualize the effects of a variance stabilizing transform as below:

```{r}
ntd <- normTransform(dds)
library("vsn")
meanSdPlot(assay(ntd))
```

If we plot the same relationship after applying the VST:

```{r}
meanSdPlot(assay(vsd))
```

DESeq also provides a slighly different method to do this transformation, 
the regularized log *rlog* method. Rougly, the benefit of this method is that the values are 
more comparable to the standard log-transformed expression values. However, as you can
see it is more strict with the lowly expressed genes. 

```{r}
rld <- rlog(dds)
meanSdPlot(assay(rld))
```


### Principal components plot

The VST data is appropriate for calculating distances between samples
or for performing PCA. More information about PCA and distance
calculation can be found in the
[RNA-seq gene-level workflow](https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#deseq2-import-functions).
In short, PCA plots allow us to visualize the most dominant axes of
variation in our data, which is useful for both quality control, and
to get a sense of how large the inter-sample differences are across
and within conditions. Here we see that PC1 (the primary axis of
variation in the data) separates the treated and untreated samples:

```{r}
plotPCA(vsd, "dex")
```

With some additional *ggplot2* code, we can also indicate which
samples belong to which cell line:

```{r}
library("ggplot2")
pcaData <- plotPCA(vsd, intgroup = c( "dex", "cell"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(x = PC1, y = PC2, color = dex, shape = cell)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()
```

Note that we do not recommend working with the transformed data for
the primary differential expression analysis. Instead we will use the
original counts and a *generalized linear model* (GLM) which takes
into account the expected variance from either low or high counts.
For statistical details, please refer to the *DESeq2* methods paper
[@Love2014Moderated].

### Choosing the proper design formula

As we saw in the plots above, while the samples are well split by treatment status,
they also segregate well by cell line type. In this case, it would be beneficial to 
explicitly control for cell line when doing our comparison between treated and 
untreated, to make sure the differentially expressed genes come as a result of the treatment.


We can set the design formula to group our samples by the two variables, 
cell and dex. We want to control for the cell line, while
testing for differences across dexamethasone treatment, so we use a
design of `~ cell + dex`, with the variable of interest to compare last 
(the last variable is the default DESeq does differential expression for).
Notice that we are creating a new DESeqDataSet here, to erase any of the 
calculations done above on the improper design formula. 

```{r}
dds <- DESeqDataSet(dds, design = ~ cell + dex)
```


## Differential expression analysis

### Standard DE steps

Differential expression analysis in *DESeq2* is performed by calling
the following two functions:

```{r}
dds <- DESeq(dds)
res <- results(dds)
```

The results table `res` contains the results for each gene (in the
same order as in the *DESeqDataSet*). If we want to see the top genes,
we can order it like so:

```{r}
head(res[order(res$pvalue),])
```

We can plot the counts for the top gene using `plotCounts`:

```{r}
plotCounts(dds, which.min(res$pvalue), "dex")
```

We can examine all the log2 fold changes (LFC) due to dexamethasone
treatment over the mean of counts using `plotMA`:

```{r}
plotMA(res, ylim=c(-5,5))
```

Note that there are many large LFC which are not significant (grey
points) on the left side of the MA-plot above. These obtain a large
LFC because of the imprecision of log counts. For more informative
visualization and more accurate ranking of genes by effect size (the
log fold change may sometimes be referred to as an *effect size*), we
recommend to use *DESeq2*'s functionality for shrinking LFCs. Our most
recent methodological development is the *apeglm* shrinkage estimator,
which is available in *DESeq2*'s *lfcShrink* function:

```{r}
library("apeglm")
resultsNames(dds)
res2 <- lfcShrink(dds, coef="dex_trt_vs_untrt", type="apeglm")
```

```{r}
par(mfrow=c(1,2))
plotMA(res, ylim=c(-3,3), main="No shrinkage")
plotMA(res2, ylim=c(-3,3), main="apeglm")
```

### Minimum effect size

If we don't want to report as significant genes with small LFC, we can
specify a minimum *biologically meaningful* effect size, by choosing
an LFC and testing against this. We can either perform such a
threshold test using the unshrunken LFCs or the LFCs provided by
*lfcShrink* using the *apeglm* method:

```{r}
res.lfc <- results(dds, lfcThreshold=1)
res.lfc2 <- lfcShrink(dds, coef="dex_trt_vs_untrt", type="apeglm",
                      lfcThreshold=1)
```

Note that *testing* against an LFC threshold is not equivalent to
testing against a null hypothesis of 0 and then filtering on LFC
values. We prefer the former, as discussed in @Love2014Moderated and
@Zhu2018. 

The *apeglm* method provides s-values [@Stephens2016] when
`svalue=TRUE` or when we supply a minimum effect size as above. These
are analogous to q-values or adjusted p-values, in that the genes with
s-values less than $\alpha$ should have an aggregate rate of false
sign or being smaller in absolute value than our given LFC threshold,
which is bounded by $\alpha$.

```{r}
par(mfrow=c(1,2))
plotMA(res.lfc, ylim=c(-5,5), main="No shrinkage, LFC test")
plotMA(res.lfc2, ylim=c(-5,5), main="apeglm, LFC test", alpha=0.01)
```

## *AnnotationHub*

### Querying *AnnotationHub*

We will use the *AnnotationHub* package to attach additional
information to the results table. *AnnotationHub* provides an
easy-to-use interface to more than 40,000 annotation records. A
record may be peaks from a ChIP-seq experiment from ENCODE, the
sequence of the human genome, a *TxDb* containing information about
transcripts and genes, or an *OrgDb* containing general information
about biological identifiers for a particular organism.

```{r}
library("AnnotationHub")
ah <- AnnotationHub()
```

The following code chunk, un-evaluated here, launches a browser for
navigating all the records available through *AnnotationHub*.

```{r eval=FALSE}
display(ah)
```

We can also query using keywords with the *query* function:

```{r}
query(ah, c("OrgDb","Homo sapiens"))
```

To pull down a particular record we use double brackets and the *name*
of the record:

```{r}
hs <- ah[["AH66156"]]
hs
```

### Mapping IDs

The *rownames* of the results table are Ensembl IDs, and most of these
are entries in *OrgDb* (although thousands are not).

```{r}
columns(hs)
table(gsub(rownames(res), pat="\\.[0-9]+$", rep="") %in% keys(hs, "ENSEMBL"))
```

We can use the *mapIds* function to add gene symbols, using `ENSEMBL`
as the keytype, and requesting the column `SYMBOL`. 

```{r}
res$symbol <- mapIds(hs, gsub(rownames(res), pat="\\.[0-9]+$", rep=""), column="SYMBOL", keytype="ENSEMBL")
head(res)
```

## Building reports

### *ReportingTools*

There are many packages for building interactive reports from
Bioconductor. Two of these are *ReportingTools* and *Glimma*, which
both provide HTML reports that allow for collaborators to examine the
top genes (or whatever features of interest) from a genomic analysis.

The code for compiling a *ReportingTools* report is:

```{r}
library("ReportingTools")
tmp <- tempdir() # you would instead use a meaningful path here
rep <- HTMLReport(shortName="airway", title="Airway DGE",
                  basePath=tmp, reportDirectory="report")
publish(res, rep, dds, n=20, make.plots=TRUE, factor=dds$dex)
finish(rep)
```

This last line, un-evaluated would launch the report in a web browser:

```{r eval=FALSE}
browseURL(file.path(tmp,"report","airway.html"))
```

### *Glimma*

Another package which can generate interactive reports is
*Glimma*. The *glMDPlot* constructs an interactive MA-plot where
hovering over a gene in the MA-plot on the left side will display the
counts for the samples on the right hand side. Clicking will bring up
the gene's information in a tooltip and in a list at the bottom of the
screen. Hovering on a sample on the right hand side will give the
sample ID in a tooltip.

```{r}
library("Glimma")
status <- as.numeric(res$padj < .1)
anno <- data.frame(GeneID=rownames(res), symbol=res$symbol)
glMDPlot(res2, status=status, counts=counts(dds,normalized=TRUE),
         groups=dds$dex, transform=FALSE,
         samples=colnames(dds), anno=anno,
         path=tmp, folder="glimma", launch=FALSE)
```

This last line would launch the report in a web browser:

```{r eval=FALSE}
browseURL(file.path(tmp,"glimma","MD-Plot.html"))
```
